client/client-multi/client-cpp/impl/utils/ThreadPool.cpp

```diff
@@ -16,89 +16,94 @@
-{
-    KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPool_.threadPoolGuard_);
-
-    while (threadPool_.isRun_) {
-        while (threadPool_.isRun_ && threadPool_.tasks_.empty() && !threadPool_.isPendingShutdown_) {
-            KAA_CONDITION_WAIT(threadPool_.onNewTask_, tasksLock);
-        }
+    void operator() ()
+    {
+        while (true) {
+            ThreadPoolTask task;
 
-        if (!threadPool_.isRun_ || (threadPool_.isPendingShutdown_ && threadPool_.tasks_.empty())) {
-            return;
-        }
+            {
+                KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPool_.threadPoolGuard_);
 
-        auto task = threadPool_.tasks_.front();
-        threadPool_.tasks_.pop_front();
+                threadPool_.onNewTask_.wait(tasksLock,
+                        [this]{ return threadPool_.state_ == ThreadPool::State::STOPPED || !threadPool_.tasks_.empty(); });
 
-        KAA_UNLOCK(tasksLock);
+                if (threadPool_.state_ == ThreadPool::State::STOPPED && threadPool_.tasks_.empty()) {
+                    return;
+                }
 
-        try {
-            task();
-        } catch (...) {}
+                task = std::move(threadPool_.tasks_.front());
+                threadPool_.tasks_.pop_front();
+            }
 
-        KAA_LOCK(tasksLock);
+            try {
+                task();
+            } catch (...) {}
+        }
     }
-}
 
-ThreadPool::ThreadPool(std::size_t workerCount): workerCount_(workerCount)
+private:
+    ThreadPool& threadPool_;
+};
+
+ThreadPool::ThreadPool(std::size_t workerCount)
+    : workerCount_(workerCount)
 {
     if (!workerCount_) {
-        throw KaaException(boost::format("Failed to create thread pool with %u workers ") % workerCount_);
+        throw std::invalid_argument((boost::format("Wrong thread pool worker count %u") % workerCount_).str());
     }
 }
 
 ThreadPool::~ThreadPool()
 {
-    stop(!shutdownTimer_);
+    stop(true);
+
+    for (auto &worker : workers_) {
+        if (worker.joinable()) {
+            worker.join();
+        }
+    }
 }
 
 void ThreadPool::add(const ThreadPoolTask& task)
 {
     if (!task) {
-        throw KaaException("Failed to add task to thread pool: empty callback");
+        throw std::invalid_argument("Null thread pool task");
     }
-    KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPoolGuard_);
 
-    if (isPendingShutdown_) {
-        throw KaaException("Failed to add task to thread pool: pending shutdown");
-    }
+    {
+        KAA_MUTEX_UNIQUE_DECLARE(tasksLock, threadPoolGuard_);
 
-    if (workers_.empty()) {
-        start();
-    }
+        if (state_ == State::CREATED) {
+            start();
+        }
 
-    tasks_.push_back(task);
+        if (state_ != State::RUNNING) {
+            throw std::runtime_error("Thread pool pending shutdown");
+        }
 
-    KAA_UNLOCK(tasksLock);
+        tasks_.push_back(task);
+    }
 
     onNewTask_.notify_one();
 }
 
 void ThreadPool::awaitTermination(std::size_t seconds)
```

Comment:
The implementation of this class is almost similar to Java ExecutorService
(see https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html)